tingbot/cache.py

```diff
@@ -0,0 +1,135 @@
+#image cache
+import time
+import re
+import rfc822
+import calendar
+import requests
+import io
+import os
+from urlparse import urlparse
+
+def get_http_timestamp(dt):
+    return calendar.timegm(rfc822.parsedate(dt))
+
+def get_last_modified(response):
+    """try and determine the last_modified time of a request"""
+    try:
+        return get_http_timestamp(response.headers['last-modified'])
+    except (KeyError,TypeError):
+        return None
+
+def get_max_age(response,last_modified):
+    """return how many seconds from original access this response is valid for"""
+    try:
+        return int(re.match(r"max-age\W*=\W*(\d+)", response.headers['cache-control']).group(1))
+    except (KeyError,ValueError,AttributeError,TypeError) as e:
+        pass
+    try:
+        return int(response.headers['max-age'])
+    except (KeyError,ValueError):
+        pass
+    try:
+        return get_http_timestamp(response.headers['expires']) - time.time()
+    except (KeyError,TypeError):
+        #really no info from server  so guess based on last-modified
+        if last_modified:
+            return min(24*60*60,(time.time()-last_modified)/10)
+        else:
+            #not even a last_modified - so conservative guess of 60s
+            return 60
+
+def get_etag(response):
+    try:
+        return response.headers['etag']
+    except KeyError:
+        return None
+
+def is_url(loc):
+    """returns true if loc is a url, and false if not"""
+    return (urlparse(loc).scheme != '')
+
+
+class ImageEntry(object):
+    #abstract base class
+    def get_image(self):
+        self.last_accessed = time.time()
+        return self.image  
+        
+    def get_size(self):
+        return self.image.surface.get_buffer().length  
```

Comment:
This will only count 1 frame for GIFs - since the `surface` variable changes depending on the time. I think this would be better as a property on Image and GIFImage classes